File: SourceCode/JavaScript/Sorting/Bubblesort.js
function sorttime(B) {
  var i;
  var A = [];
  A.length = B.length;
  var totaltime, runs;
  var numruns = 20;

  for(i = 0; i < B.length; i++) A[i] = B[i];
  time1 = new Date();
  bubblesort(A);
  time2 = new Date();
  checkorder(A);
  console.log("Standard Bubble Sort: Size " + testsize + ", Time: " + (time2 - time1));

  for(i = 0; i < B.length; i++) A[i] = B[i];
  time1 = new Date();
  bubblesortcheck(A);
  time2 = new Date();
  checkorder(A);
  console.log("Swap Check Bubble Sort: Size " + testsize + ", Time: " + (time2 -time1));

  for(i = 0; i < B.length; i++) A[i] = B[i];
  time1 = new Date();
  bubblesortcheck2(A);
  time2 = new Date();
  checkorder(A);
  console.log("Swap Check Bubble Sort 2: Size " + testsize + ", Time: " + (time2 - time1));

totaltime = 0;
for (runs=0; runs<numruns; runs++) {
  for(i = 0; i < B.length; i++) A[i] = B[i];
  time1 = new Date();
  wikipedia(A);
  time2 = new Date();
  checkorder(A);
totaltime += (time2-time1);
}
  console.log("Wikipedia Bubble Sort: Size " + testsize + ", " + " for " + numruns + " runs, Time: " + totaltime);

  for(i = 0; i < B.length; i++) A[i] = B[i];
  time1 = new Date();
  wikipedia2(A);
  time2 = new Date();
  checkorder(A);
  console.log("Wikipedia Bubble Sort 2: Size " + testsize + ", Time: " + (time2 - time1));

totaltime = 0;
for (runs=0; runs<numruns; runs++) {
  for(i = 0; i < B.length; i++) A[i] = B[i];
  time1 = new Date();
  unwikipedia(A);
  time2 = new Date();
  checkorder(A);
totaltime += (time2-time1);
}
  console.log("Wikipedia Bubble Sort without swap checking: Size " + testsize + ", " + " for " + numruns + " runs, Time: " + totaltime);
}

// A flag check if a pass did not have any swaps, which lets us quit
function bubblesortcheck(A) {
  for (var i = 0; i < A.length - 1; i++) { // Insert i'th record
    var swaps = false;
    for (var j = 1; j < A.length - i; j++)
      if (A[j-1] > A[j]) {
        swap(A, j-1, j);
        swaps = true;
      }
    if(!swaps) { console.log("Quit at " + i); break; }  // Can quit early
  }
}

// Modify the flag to check position of last swap taken
function bubblesortcheck2(A) {
  for (var i = 0; i < A.length - 1; i++) {// Insert i'th record
    var lastseen = 0;
    var top = A.length;
    for (var j = 1; j < top; j++)
      if (A[j - 1] > A[j]) {
        swap(A, j - 1, j);
        lastseen = j - 1;
      }
    top = lastseen;
    if (top == 0) { console.log("Quit at " + i); break; }  // Can quit early
  }
}

// Wikipedia article "optimization" to only swap up to the last swap seen
function wikipedia(A) {
  var n = A.length - 1;
  while (n > 0) {
    var newn = 0;
    for (var i = 0; i < n; i++)
      /* if this pair is out of order */
      if (A[i] > A[i + 1]) {
        swap(A, i, i + 1);
        newn = i;
      }
    n = newn;
  }
}

// Wikipedia article "optimization" rewritten with a for loop
function wikipedia2(A) {
  var newn;
  for(var n = A.length - 1; n > 0; n = newn) {
    newn = 0;
    for (var i = 0; i < n; i++)
      /* if this pair is out of order */
      if (A[i] > A[i + 1]) {
        swap(A, i, i + 1);
        newn = i;
      }
  }
}

// Wikipedia article-compatible version without swap checking
function unwikipedia(A) {
  var n = A.length - 1;
  while (n > 0) {
    for (var i = 0; i < n; i++)
      /* if this pair is out of order */
      if (A[i] > A[i + 1]) {
        swap(A, i, i + 1);
      }
    n -= 1;
  }
}

function success() {
  var bw = require("buffered-writer");
  bw.open("BubblesortTestSuccess").write("Success").close();
}

function sorttest(A) {
  bubblesort(A);
}

/* *** ODSATag: Bubblesort *** */
function bubblesort(A) {
  for (var i = 0; i < A.length - 1; i++) // Insert i'th record
    for (var j = 1; j < A.length - i; j++)
      if (A[j - 1] > A[j])
        swap(A, j - 1, j);
}
/* *** ODSAendTag: Bubblesort *** */

File: SourceCode/Python/Sorting/Bubblesort.py
def sorttime(B):
  A = [randrange(1,1000) for _ in range(testsize)]  # To make it create a real A for copying
  numruns = 5

  for i in range(len(B)):
    A[i] = B[i]
  time1 = datetime.now()
  bubblesort(A)
  time2 = datetime.now()
  checkorder(A)
  print "Standard Bubble Sort: Size ", testsize, ", Time: ", millis(time2 - time1)

  for i in range(len(B)):
    A[i] = B[i]
  time1 = datetime.now()
  bubblesortcheck(A)
  time2 = datetime.now()
  checkorder(A)
  print "Swap Check Bubble Sort: Size ", testsize, ", Time: ", millis(time2 - time1)

  for i in range(len(B)):
    A[i] = B[i]
  time1 = datetime.now()
  bubblesortcheck2(A)
  time2 = datetime.now()
  checkorder(A)
  print "Swap Check Bubble Sort 2: Size ", testsize, ", Time: ", millis(time2 - time1)

  totaltime = 0
  for runs in range(1, numruns):
    for i in range(len(B)):
      A[i] = B[i]
    time1 = datetime.now()
    wikipedia(A)
    time2 = datetime.now()
    checkorder(A)
    totaltime += millis(time2-time1)
  print "Wikipedia Bubble Sort: Size ", testsize, ", for ", numruns, " runs, Time: ", totaltime

  totaltime = 0
  for runs in range(1, numruns):
    for i in range(len(B)):
      A[i] = B[i]
    time1 = datetime.now()
    unwikipedia(A)
    time2 = datetime.now()
    checkorder(A)
    totaltime += millis(time2-time1)
  print "Wikipedia-compatible Bubble Sort without swap checking: Size ", testsize, " for ", numruns, "runs, Time: ", totaltime

# A flag check if a pass did not have any swaps, which lets us quit
def bubblesortcheck(A):
  for i in range(len(A)): # Insert i'th record
    swaps = False
    for j in range (1, len(A) - i):
      if (A[j-1] > A[j]):
        swap(A, j-1, j)
        swaps = True
    if (not swaps):
      print "Quit at ", i
      break  # Can quit early

# Modify the flag to check position of last swap taken
def bubblesortcheck2(A):
  for i in range(len(A)-1): # Insert i'th record
    lastseen = 0;
    top = len(A)
    for j in range(1, top-1):
      if (A[j - 1] > A[j]):
        swap(A, j - 1, j)
        lastseen = j - 1
    top = lastseen
    if (top == 0):
      print "Quit at ", i
      break   # Can quit early

# Wikipedia article "optimization" to only swap up to the last swap seen
def wikipedia(A):
  n = len(A) - 1;
  while (n > 0):
    newn = 0
    for i in range(n):
      # if this pair is out of order
      if (A[i] > A[i + 1]):
        swap(A, i, i + 1)
        newn = i
    n = newn

# Wikipedia article-compatible version without swap checking
def unwikipedia(A):
  n = len(A) - 1;
  while (n > 0):
    for i in range(n):
      # if this pair is out of order
      if (A[i] > A[i + 1]):
        swap(A, i, i + 1)
    n -= 1

def success():
  print "Success! (Need to define this)"

def sorttest(A):
  bubblesort(A)

#/* *** ODSATag: Bubblesort *** */
def bubblesort(A):
  for i in range (len(A)): # Insert i'th record
    for j in range(1, len(A) - i):
      if (A[j - 1] > A[j]):
        swap(A, j - 1, j)
#/* *** ODSAendTag: Bubblesort *** */
